### PR TITLE
 hygiene: Decouple transparencies from expansion IDs

### LINKED ISSUE #50504: proc_macro! expansion can refer to items defined in the root w/o importing them
Apologies if this has already been reported; I didn't search too hard.

### STR

``` rust
// bar/src/lib.rs

#![feature(proc_macro)]

extern crate proc_macro;
#[macro_use]
extern crate quote;

use proc_macro::TokenStream;

#[proc_macro]
pub fn bar(_: TokenStream) -> TokenStream {
    quote!(mod __bar {
        // NOTE `Foo` is never imported in this module
        static mut BAR: Option<Foo> = None;
    }).into()
}
```

``` rust
// foo/src/lib.rs

#![feature(proc_macro)]

extern crate bar;

use bar::bar;

struct Foo;

bar!();
```

This code compiles but I would expect it not to because the expanded form doesn't compile:

``` rust
// foo/src/lib.rs -- `cargo expand`-ed and cleaned up
struct Foo;

mod __bar {
    static mut BAR: Option<Foo> = None;
}
```

``` console
$ cargo build
error[E0412]: cannot find type `Foo` in this scope
 --> src/lib.rs:4:28
  |
4 |     static mut BAR: Option<Foo> = None;
  |                            ^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
  |
4 |     use Foo;
  |

error: aborting due to previous error
```

### Meta

```
$ rustc -V
rustc 1.27.0-nightly (428ea5f6b 2018-05-06)
```

cc @jseyfried

### PR DESCRIPTION
**Pull request by petrochenkov**
And remove fallback to parent modules during resolution of names in scope.

This is a breaking change for users of unstable macros 2.0 (both procedural and declarative), code like this:
```rust
#![feature(decl_macro)]

macro m($S: ident) {
    struct $S;
    mod m {
        type A = $S;
    }
}

fn main() {
    m!(S);
}
```
or equivalent
```rust
#![feature(decl_macro)]

macro m($S: ident) {
    mod m {
        type A = $S;
    }
}

fn main() {
    struct S;
    m!(S);
}
```
stops working due to module boundaries being properly enforced.

For proc macro derives this is still reported as a compatibility warning to give `actix_derive`, `diesel_derives` and `palette_derive` time to fix their issues.

Fixes https://github.com/rust-lang/rust/issues/50504 in accordance with [this comment](https://github.com/rust-lang/rust/issues/50504#issuecomment-399764767).

### CODE REVIEW

**Comment by petrochenkov** [comment]
Needs check-only crater run (stable derives are affected).
ping @rust-lang/infra

File: src/libsyntax_pos/hygiene.rs [hunk 1]
```diff
@@ -284,22 +285,31 @@ impl SyntaxContext {
         })
     }
 
-    /// Extend a syntax context with a given mark
```
Comment (oli-obk) [inline src/libsyntax_pos/hygiene.rs]: you lost a doc comment

File: src/test/ui-fulldeps/proc-macro/auxiliary/generate-mod.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#![feature(proc_macro)]
+#![crate_type = "proc-macro"]
+
+extern crate proc_macro;
+use proc_macro::*;
+
+#[proc_macro]
+pub fn check(_: TokenStream) -> TokenStream {
+    "
+    struct Outer;
```
Comment (oli-obk) [inline src/test/ui-fulldeps/proc-macro/auxiliary/generate-mod.rs]: can you check that we can still access things outside the expansion here?

File: src/test/ui-fulldeps/proc-macro/auxiliary/generate-mod.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    mod inner {
+        type Inner = Outer; // `Outer` shouldn't be available from here
```
Comment (oli-obk) [inline src/test/ui-fulldeps/proc-macro/auxiliary/generate-mod.rs]: and that we cannot access anything from outside the expansion without an import

**Comment by kennytm** [comment]
cc @pietroalbini check-only crater run wanted (details are in the crater status spreadsheet already).

**Comment by pietroalbini** [comment]
Crater run started. It should finish in ~2 days.

**Comment by pietroalbini** [comment]
Hi @petrochenkov (crater requester), @oli-obk (PR reviewer)! Crater results are at: http://cargobomb-reports.s3.amazonaws.com/pr-51952/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by petrochenkov** [comment]
`actix_derive`, `diesel_derives` and `palette_derive` are root causes for most of regressions and they actually have bugs and use unaccessible names from parent modules.
(There are couple of other root regressions, but they use unstable proc macros.)
cc @fafhrd91 @sgrif @Ogeon @alexcrichton 

The common pattern is
```rust
#[derive(Something)]
struct MyStruct;
```
being expanded into
```rust
struct MyStruct;

mod __implementation_details {
    // Note, that `MyStruct` is not in scope here.
    impl Something for MyStruct {
        let x: SomethingElseFromTheParentModule = ...;
    }
}
```

The "universal solution" is to do this:
```rust
mod __implementation_details {

    use super::*; // <---

    // Now `MyStruct` and `SomethingElseFromTheParentModule ` are in scope.
    impl Something for MyStruct {
       let x: SomethingElseFromTheParentModule = ...;
    }
}
```

Perhaps we can send fixes, release minor versions and land this fix.

**Comment by petrochenkov** [comment]
@alexcrichton 
>Out of curiosity, is it possible to apply this fix for just proc_macro and proc_macro_attribute?

This should be possible to do with a targeted hack, if I'm not mistaken about the root cause of the bug.
I'll try tomorrow.

**Comment by petrochenkov** [comment]
Updated with the issue fixed for all macros except for proc macro derives, in which it's still reported, but as a warn-by-default compatibility lint.

**Comment by alexcrichton** [comment]
@bors: try

Nice! Let's run this through crater again to double check not too much breaks

**Comment by petrochenkov** [comment]
ping @pietroalbini
check-only crater run is needed

**Comment by pietroalbini** [comment]
Check-only crater run started on crater-2. It should finish in ~1 day.

**Comment by pietroalbini** [comment]
> **New thing!** ignore the new "errors" section, those are internal Crater errors. Before they were ignored, but now they are explicit so it's possible to fix them. They're not related to this PR though!

Hi @petrochenkov  (crater requester), @oli-obk (PR reviewer)! Crater results are at: https://cargobomb-reports.s3.amazonaws.com/pr-51952-2/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by petrochenkov** [comment]
Three non-spurious regressions:
- [`diesel_geometry-1.2.2`](https://cargobomb-reports.s3.amazonaws.com/pr-51952-2/974ece59a1c76f424fe80f13996a4acd44e0ccf7-alt/reg/diesel_geometry-1.2.2/log.txt)
- [`mocktopus-0.4.0`](https://cargobomb-reports.s3.amazonaws.com/pr-51952-2/974ece59a1c76f424fe80f13996a4acd44e0ccf7-alt/reg/mocktopus-0.4.0/log.txt)
- [`robin-0.3.0`](https://cargobomb-reports.s3.amazonaws.com/pr-51952-2/974ece59a1c76f424fe80f13996a4acd44e0ccf7-alt/reg/robin-0.3.0/log.txt)

I'll take a look today.

**Comment by Ogeon** [comment]
So, is this going to introduce a lint for the inaccessible name thing? Also, will the function solution in https://github.com/rust-lang/rust/pull/51952#issuecomment-403175646 work as a workaround? I could probably release a patch for `palette_derive`, at least, to make it available when this is released.

**Comment by alexcrichton** [comment]
@Ogeon indeed! You should be able to apply a fix with stable Rust today though, detailed by @petrochenkov [above](https://github.com/rust-lang/rust/pull/51952#issuecomment-403174029). To verify though that the fix works you may want to wait for this to land in nightly to test it out!

**Comment by Ogeon** [comment]
Alright, then there's no hurry right now. I somehow got the impression that it would be a patch or at least show up in an earlier stable. I'll start by opening an issue on my end, so it's not forgotten, and see how this develops.

**Comment by alexcrichton** [comment]
@bors: r+

Ok great!

**Comment by sgrif** [comment]
So this completely breaks Diesel and I don't see an obvious way to migrate us. Specifically our derives have to satisfy:

- May be used on structs inside a function as well as a normal scope
- May be used inside of Diesel itself

This means that we can't actually assume that `extern crate diesel` works, so we've had to resort to the following:

```rust
// Inside of Diesel:
#[macro_export]
macro_rules! __diesel_use_everything! {
    () => { pub use $crate::*: }
}

// In the derive
mod __dummy_mod {
    extern crate std;

    mod diesel {
        __diesel_use_everything!()
    }
}
```

None of the solutions proposed will work here. Adding `use super::*` means that our derives no longer work on structs defined in functions. Changing to use a function instead of a module means we can no longer reliably import items from Diesel.

**Comment by alexcrichton** [comment]
@sgrif is the basic problem that you basically need `::diesel::...` to work everywhere but it doesn't work within the diesel crate itself? Can that be solved with a [trick like this](https://github.com/rustwasm/wasm-bindgen/blob/ed05c7b9454905f15488714619950a55cd3c8d2d/src/lib.rs#L52-L54)?

**Comment by sgrif** [comment]
I'll give that a shot. It still means that we require `extern crate diesel` at the root, which is an unfortunate superpower that built-in derives have which other crates have no way to replicate.

**Comment by alexcrichton** [comment]
Yes that's an unfortunate hygiene issue where you really want the definition of the macro to say "please define what this crate should resolve to", but we don't have a way to do that today. For now it matches most users' expectations to put `diesel` in their Cargo.toml and the end goal is that's all they put in their manifest as they'll be importing macros and runtime values from that one same crate. (and eventually there's no need for `extern crate diesel` as well)

**Comment by sgrif** [comment]
I agree it's unlikely that the crate is being renamed, but it would be great to ensure that derives from crates have all the same capabilities as built-in derives in the future -- For the time being your suggestion does seem to work though https://github.com/diesel-rs/diesel/pull/1787

**Comment by sgrif** [comment]
@alexcrichton I've opened #52545 which is related to this. It sounds like the fix for that is not straightforward or easy. Currently there's no obvious migration path for crates which are affected by this. What do you think about reverting this for the time being until there's a clear path forward for existing crates?

**Comment by sgrif** [comment]
I'm also surprised that this change was made so quietly. This is a pretty major breaking change to a feature that's been stable for a year and a half. I would like to think this would have at least received an RFC. This is very clearly not a soundness issue, and whether it is even a bug is arguable IMO.

This makes me extremely worried about whether Rust will actually live up to its stability promises. For a feature that has been stable for this long, where there were at least a dozen *known* cases of folks relying on this behavior, there really should have been more discussion -- not just on whether the behavior was intended or not, but on whether it's reasonable to change it at this point given that Rust is meant to be stable.

**Comment by petrochenkov** [comment]
We should not revert this PR as is and reintroduce the issues that were fixed here, that would be harmful for our hygiene system long-term.
The special compatibility code that is currently supported under lint is enough to cover diesel's use case for some time.
I think we can tweak the lint so it can be `allow`-ed from inside the diesel at most (currently it's attached to crate root and can't be disabled locally, but we can do better).

**Comment by petrochenkov** [comment]
These are all consequences of rushed stabilization of macros 1.2 and low "bus factor" for hygiene system (not enough qualified reviewing was done).

**Comment by sgrif** [comment]
While I appreciate where you're coming from, stability should not mean "unless we decide we change our minds". I agree with you that this is something that should have been discussed before the feature became stable in the first place, but that does not mean that it should be OK for the rug to get pulled out from under the ecosystem on a stable feature.

**Comment by alexcrichton** [comment]
@sgrif did diesel or any other crates you know of stop compiling as a result of this change? This change was landed under the presumption that nothing was broken and crates continued to compile, if that's not the case we should fix that!

If crates continued to compile, however, then I think it's a bit disingenous to say that the rug was pulled out from under you or that this is a major breaking change. This [fixed bug](https://github.com/rust-lang/rust/issues/50504) exhibits behavior which is clearly a bug in procedural macros which [has affected stable Rust](https://github.com/rust-lang/rust/issues/50504#issuecomment-387434157) for quite some time.

We always try quite hard to keep crates compiling on stable. This does not mean that we will prevent ourselves from fixing bugs though as we're doing in this situation. We are tasked, however, with fixing the bug while keeping existing code compiling.

**Comment by petrochenkov** [comment]
@sgrif 
Bugs with compatibility implications regularly happened and will happen in the future. A lot of them are fixed silently and nobody notices, and for those with noticeable impact the process so far was to continue support, but issue a deprecation warning - exactly what was done here.

**Comment by sgrif** [comment]
@alexcrichton 

> did diesel or any other crates you know of stop compiling as a result of this change? 

Yes, Diesel did. You also listed a ton of affected crates in https://github.com/rust-lang/rust/issues/50504#issuecomment-389734217. Again, I think calling it a "fixed bug" is quite arguable. Personally I'd expect hygiene to mean that a given ident is resolved in the scope it was given from. My main point is that I don't think it's an obvious conclusion one way or the other, and I would have liked to see a discussion around it.

> We always try quite hard to keep crates compiling on stable. This does not mean that we will prevent ourselves from fixing bugs though as we're doing in this situation.

I agree. But there does become a point when folks start relying on behavior enough that there is a conversation to be had about whether fixing the "bug" does more harm than good (there are several cases like this in Rails for example). At a certain point something in the grey area ends up being a defacto feature.

I think it's also important to define what "bug" means here. I couldn't find anything in the relevant RFCs which explicitly defines how this behavior should interact. At best this is a clarification of intent, but given that it's on a stable feature, I don't think that should be done without giving the community an opportunity to voice their opinion.

@petrochenkov 

> Bugs with compatibility implications regularly happened and will happen in the future. 

I agree. I do not think this is obviously a bug. It's also much more nuanced than you are making it out to be, see my above comments.

**Comment by sgrif** [comment]
@alexcrichton 

>  so to be clear, diesel compiled on stable, and then the exact same source code no longer compiles on nightly? I'm not talking about warnings, but literal compiler errors.

Most published versions of Diesel were shipped with `#[deny(warnings)]` -- A practice that we have since stopped. That said, this is still going to eventually become an error. And regardless of whether something is an error or a warning, users will still report compiler warnings as bugs to crate maintainers, making it basically the same from our point of view.

> The crater report you gisted is not even tied to this change, so I'd just want to be clear about what breaking is.

I linked to a discussion on the issue closed by this PR, not sure how it's unrelated. Pretty much every crate there is generating a module for the same reason -- to be able to have an isolated scope for `use`, and to be able to safely add `extern crate` to work around the lack of `$crate`.

> I agree there's possibility for a gray area, but I disagree that this is not obviously a bug. All of macros 1.1 is "call site" hygiene which is "as if the tokens are copy/pasted". That's clearly not the case when you're generating mod statements that don't actually import the items inside of them.

This is a fair point. Really there's a broader issue here is that proc macros (and derives in particular) have no way to isolate their scope, and want to do things that don't have obvious interactions with hygiene (such as `use` statements). This is part of why I wish there had been a bit more discussion. Rather than the use case folks are relying on being considered, warnings/errors being generated for our users, and no workable path forward.

**Comment by petrochenkov** [comment]
>All of macros 1.1 is ... "as if the tokens are copy/pasted".

I think this is what was specified in the RFC for 1.1 and announcement for 1.2, yes.
In other words, macros 1.1-1.2 don't provide "qualitatively new" abilities compared to non-macro code, e.g. just writing tokens by hand. (Compare with `macro_rules` that exhibit the same "copy-pasting" behavior and don't allow this trick.)

**Comment by alexcrichton** [comment]
@sgrif our stability policy does not consider new warnings or changes to warnings a breaking change due to `#[deny(warnings)]`. That's because it doesn't affect downstream usage of a crate, only the crate itself. That should mean that all crates depending on `diesel` continue to work, despite `diesel` itself needing a fix (which can be immediately remediated by removing `deny(warnings)` for the near-term).

The crater report you linked was a blanket denial of all procedural macros generating `mod` statements. This PR is not that change. This PR is a change that warns about using items that weren't imported in generated modules. While that report you linked may be similar to the fallout here, it is only tangentially related.

There is many many issues with procedural macros and how they do not have many hygiene controls, they're forced to just splat tokens into the user's program. These problems are well-known though and the solution is under development but is still not ready for stabilization.

**Comment by sgrif** [comment]
> our stability policy does not consider new warnings or changes to warnings a breaking change due to #[deny(warnings)]. 

I understand that. This is not just a new default warn lint. This is a warning for what will end up being a breaking change. It also doesn't really matter in terms of the impact that this has on downstream crates. If someones build starts spamming warnings, we still get multiple reports a day (the most recent versions of Diesel do not `#[deny(warnings)]`, but we have still gotten multiple reports about this on those versions)

> The crater report you linked was a blanket denial of all procedural macros generating mod statements. This PR is not that change.

I looked through about half of them, and every single one I looked at is going to be affected by this change. As I said in my previous comment, there is very little reason to emit a module from a `derive` unless you are using it to isolate scope.

> These problems are well-known though and the solution is under development but is still not ready for stabilization.

Has there been an RFC for this? This is my main point. This is a *stable* feature, and it really feels like things about it are just being changed with no input from the community.

**Comment by sgrif** [comment]
Regardless of anything else, I hope you understand the frustration here. A lot of folks use Diesel with Rocket or other crates that require nightly, so we tend to feel pains from that pretty quick. Users on older versions that still had `#[deny(warnings)]` have had their builds break completely. Others are just spammed with warnings. Both have opened issues and complained in our support channels. Today there is *no* migration path forward for us to fix this issue for our users.

**Comment by petrochenkov** [comment]
Sorry :(
I'll try to fix the lint so it can be `allow`-ed locally in diesel without bothering clients (this weekend most probably).

**Comment by alexcrichton** [comment]
@sgrif yes I'm sorry that Diesel is going through these troubles, this particular migration could probably have gone more smoothly!

The warnings here were intended to be like all other future incompatibility warnings the compiler issues: they don't break existing code (modulo lint denial) and they are supressable in the near term if necessary. The migration path here was thought to be simple enough but it wasn't exercised for Diesel's case specifically and you've discovered there's a bug or two which prevents a feasible migration path, but we'll work on that!

The compiler has a [well documented policy](https://forge.rust-lang.org/rustc-bug-fix-procedure.html) on how to fix a bug in the compiler. It appears we accidentally left out the tracking issue part of that, but otherwise that policy was followed. To reiterate some motivation of that policy, no future-incompatible compiler warning is turned into an error, ever, if enough of the ecosystem is still using it. We would never consider turning this into a hard error unless Diesel had successfully migrated to no longer have the warning issued. I disagree with your point of "is no different from it not working", this is the crucial distinction.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/52589

**Pull request merged**